invisible_cities/reco/deconv_functions.py

```diff
@@ -0,0 +1,317 @@
+                       ) -> Callable:
+    """
+    Prepares the given data for deconvolution. This involves interpolation of
+    the data.
+
+    Parameters
+    ----------
+    data        : Sensor (hits) position points.
+    weight      : Sensor charge for each point.
+
+    Initialization parameters:
+        sample_width : Sampling size of the sensors.
+        bin_size     : Size of the interpolated bins.
+        inter_method : Interpolation method.
+
+    Returns
+    ----------
+    Hs          : Charge input for deconvolution.
+    inter_points : Coordinates of the deconvolution input.
+    """
+
+    def deconvolution_input(data        : Tuple[np.ndarray, ...],
+                            weight      : np.ndarray
+                           ) -> Tuple[np.ndarray, Tuple[np.ndarray, ...]]:
+
+        ranges = [[coord.min() - 1.5 * sw, coord.max() + 1.5 * sw]     for coord,   sw in zip(data      , sample_width)]
+        nbin   = [np.ceil(np.diff(rang)/bs).astype('int')[0]           for bs   , rang in zip(bin_size  ,       ranges)]
+        if inter_method is not InterpolationMethod.none:
+            allbins = [np.linspace(*rang, np.ceil(np.diff(rang)/sw)+1) for rang ,   sw in zip(ranges[:2], sample_width)]
+            Hs, edges = np.histogramdd(data, bins=allbins, normed=False, weights=weight)
+        else:
+            Hs, edges = np.histogramdd(data, bins=nbin   , normed=False, weights=weight, range=ranges)
+
+        inter_points = np.meshgrid(*(shift_to_bin_centers(edge) for edge in edges), indexing='ij')
+        inter_points = tuple      (inter_p.flatten() for inter_p in inter_points)
+        if inter_method is not InterpolationMethod.none:
+            Hs, inter_points = interpolate_signal(Hs, inter_points, edges, nbin, inter_method)
+
+        return Hs, inter_points
+
+    return deconvolution_input
+
+
+def interpolate_signal(Hs           : np.ndarray,
+                       inter_points : Tuple[np.ndarray, ...],
+                       edges        : Tuple[np.ndarray, ...],
+                       nbin         : List[int],
+                       inter_method : InterpolationMethod = InterpolationMethod.cubic
+                       ) -> Tuple[np.ndarray, Tuple[np.ndarray, ...]]:
+    """
+    Interpolates an n-dimensional distribution along N points. Interpolation
+    has a lower limit equal to 0.
+
+    Parameters
+    ----------
+    Hs           : Distribution weights to be interpolated.
+    inter_points : Distribution coordinates to be interpolated.
+    edges        : Edges of the coordinates.
+    nbin         : Number of points to be interpolated in each dimension.
+    inter_method : Interpolation method.
+
+    Returns
+    ----------
+    H1         : Interpolated distribution weights.
+    new_points : Interpolated coordinates.
+    """
+    new_points   = np.meshgrid(*(shift_to_bin_centers(np.linspace(np.min(edge), np.max(edge), n + 1)) for n, edge in zip(nbin, edges)), indexing='ij')
```

Comment:
try to break up this line a bit. you can even define the inner tuple in a different line